File: top100/22_generate_parentheses/genParentheses.java
class Solution 
{
    public List<String> generateParenthesis(int n) 
    {
        List<String> res = new ArrayList<>();
        helper(res, "", n, n);
        return res;
    }
    
    private void helper(List<String> res, String curr, int left, int right)
    {
        if (left == 0 && right == 0)
        {
            res.add(curr);
            return;
        }
        
        if (left > 0)
            helper(res, curr+"(", left-1, right);
        
        if (right > left)
            helper(res, curr+")", left, right-1);
    }
}

/**
 *    recursion tree: https://www.youtube.com/watch?v=XF0wh8M2A6E
 *    java code: https://www.cnblogs.com/grandyang/p/4444160.html
 *               https://www.youtube.com/watch?v=ptYZDPk2bOg&t=212s                                  
 *   
 */

File: top100/22_generate_parentheses/genParentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        self.helper(res, n, n, "")
        return res
    
    def helper(self, res, left, right, curr):
        if left == 0 and right == 0:
            res.append(curr)
            return
        if left > 0:
            self.helper(res, left-1, right, curr+"(")
        if right > left:
            self.helper(res, left, right-1, curr+")")